pkg/clusterapi/aws/utils.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/aws/aws-sdk-go/service/elb"
+	"github.com/aws/aws-sdk-go/service/elb/elbiface"
+
+	cov1 "github.com/openshift/cluster-operator/pkg/apis/clusteroperator/v1alpha1"
+)
+
+// SortInstances will examine the given slice of instances and return the current active instance for
+// the machine, as well as a slice of all other instances which the caller may want to terminate. The
+// active instance is calculated as the most recently launched instance.
+// This function should only be called with running instances, not those which are stopped or
+// terminated.
+func SortInstances(instances []*ec2.Instance) (*ec2.Instance, []*ec2.Instance) {
```

Comment:
This is the approach that I like the very least. There is no reason that this function needs to return the second argument. It only adds a concern to the function that is orthogonal to what you want the function to do--which is to find the instance that is the most recent. I **highly** recommend that this function simply returns the index of the most recent instance. It is simple enough for the caller to do what it needs to do with that index.

```
newest := FindNewestInstance(instances)
instances[newest] = instances[len(instances)-1] 
instances = instances[:len(instances)-1]
```
or
```
newest := FindNewestInstance(instances)
instances = append(instances[:newest], instances[i+newest:]...)
```

Reply:
Your interpretation of what the function is meant to do is not the same as intended,  GetInstance below is for finding the instance that is most recent. I want this function to separate the active from the rest, for callers who need to act on both. I don't want a higher level caller to have to interpret positions in an array or do a bunch of slice logic especially when we're literally talking about destroying machines. The logic here is nicely isolated and very easy to unit test, and the caller gets back exactly what they need to act on.

I think we're way too far into the weeds here and I'm on about the 4th rewrite of this, we need to move this along. I am guessing your issue is with the allocation of the terminate slice unnecessarily for callers of GetInstance who literally just want the active, assuming some bizarre situations where there are thousands of instances matching the machine?

I don't want to push that slice logic up as high as the actuator because it gets harder to test, so I will add a probably unexported findNewestInstanceIndex, and use it in GetInstance. I'd like to keep SortInstances and use your implementation above and continue returning the separate arguments. I will document that you should not use sort instances for performance reasons unless you need to act on the inactive instances. Maybe rename to SortActiveFromInactive. Does this sound ok?

Reply:
No, that is not my issue. My issue is that the `SortInstance` function is trying to do too many things--the smell being the multiple returns from the function.

Your argument that the logic is nicely isolated and very easy to unit test is inaccurate. In the end, we need unit tests for `GetInstance`, which will duplicate in many ways the same test cases and results in `TestSortInstances`. If we instead had a `GetInactiveInstances` function which returned only the inactive instances, then the logic for `GetActiveInstance` and `GetInactiveInstances` would be more isolated than the logic is currently for `SortInstances`. And their corresponding unit tests would be smaller and even easier to test.

The calling code is objectively simpler to read, too
```
newestInstance := GetActiveInstances(instances)
terminateInstances := GetInactiveInstances(instances)
if len(terminateInstances) > 0 {
  err = TerminateInstances(client, terminateInstances, mLog)
  if err != nil {
    return err
  }
}
```

With that being said, these are just my opinions and not something that I would hold up the PR for. If you are satisfied with the way the functions are laid out, then that is good enough for me.

Reply:
This last proposal doesn't look bad to me, my main goal was that the caller (currently the actuator) not have to play with array indicies and assumptions. However the separating the two in one function does seem ok for now. Thanks.